src/test/java/io/cdap/plugin/batch/ETLFTPTestRun.java

```diff
@@ -0,0 +1,201 @@
+
+    Assert.assertEquals("Expected records", 1, output.size());
+    Set<String> outputValue = new HashSet<>();
+    for (StructuredRecord record : output) {
+      outputValue.add(record.get("body"));
+    }
+    Assert.assertTrue(outputValue.contains("Hello World"));
+  }
+
+  private void testHelper(Map<String, String> properties, Map<String, String> runTimeProperties) throws Exception {
+    ETLStage source = new ETLStage("source", new ETLPlugin(
+      "FTP", BatchSource.PLUGIN_TYPE, properties, null));
+
+    String outputDatasetName = "testing-ftp-source-with-macro";
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder()
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    ApplicationManager appManager = deployETL(etlConfig, "FTPWithMacro");
+    runETLOnce(appManager, runTimeProperties);
+
+    DataSetManager<Table> outputManager = getDataset(outputDatasetName);
+    List<StructuredRecord> output = MockSink.readOutput(outputManager);
+
+    Assert.assertEquals("Expected records", 1, output.size());
+    Set<String> outputValue = new HashSet<>();
+    for (StructuredRecord record : output) {
+      outputValue.add(record.get("body"));
+    }
+    Assert.assertTrue(outputValue.contains("Hello World"));
+  }
+
+  @Test
+  public void testFTPBatchSourceWithMacro() throws Exception {
+    testHelper(ImmutableMap.of("path", "${path}", "referenceName", "${referenceName}"),
+               ImmutableMap.of("path", String.format("ftp://%s:%s@localhost:%d%s",
+                                                     USER, PWD, port, folder.getAbsolutePath()),
+                               "referenceName", "ftp_with_macro"));
+  }
+
+  @Test
+  public void testFTPBatchSourceWithInvalidPath() {
+    ETLStage source = new ETLStage("source", new ETLPlugin(
+      "FTP",
+      BatchSource.PLUGIN_TYPE,
+      ImmutableMap.<String, String>builder()
+        .put(PATH, String.format("ftp://localhost:%d%s",
+                                 port, folder.getAbsolutePath()))
+        .put(IGNORE_NON_EXISTING_FOLDERS, "false")
+        .put("referenceName", "ftp")
+        .build(),
+      null));
+
+    String outputDatasetName = "testing-ftp-source";
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputDatasetName));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder()
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(DATAPIPELINE_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("FTPBatchSourceWithInvalidPath");
+
+    // deploying would thrown an exception
+    try {
+      deployApplication(appId, appRequest);
+      Assert.fail("Missing authentication in url");
+    } catch (Exception e) {
+      //
```

Comment:
can you check that the exception thrown here is the one that's expected?

Reply:
I think you can't catch the specific exception in this scenario. We can't catch _ValidationException_ like we did in here:

https://github.com/AdaptiveScale/ftp-plugins/blob/67d6b7c3502b818047ce441218a9cc00007209aa/src/test/java/io/cdap/plugin/batch/source/FTPBatchSourceTest.java#L72-L85

...because in this test we are trying to deploy the app, not checking the config. 

I've seen similar tests like this one in here for example: 

https://github.com/AdaptiveScale/hydrator-plugins/blob/7c304a30033847c55bc2990dd7068bcd0990e9e8/core-plugins/src/test/java/io/cdap/plugin/batch/ETLMapReduceTestRun.java#L57-L84

Reply:
I see, thanks.